pkg/frontend/admin_openshiftcluster_kubernetesobjects.go

```diff
@@ -0,0 +1,171 @@
+
+	err := validateAdminKubernetesObjectsNonCustomer(r.Method, kind, namespace, name)
+	if err != nil {
+		return err
+	}
+
+	resourceID := strings.TrimPrefix(r.URL.Path, "/admin")
+
+	doc, err := f.db.OpenShiftClusters.Get(ctx, resourceID)
+	switch {
+	case cosmosdb.IsErrorStatusCode(err, http.StatusNotFound):
+		return api.NewCloudError(http.StatusNotFound, api.CloudErrorCodeResourceNotFound, "", "The Resource '%s/%s' under resource group '%s' was not found.", vars["resourceType"], vars["resourceName"], vars["resourceGroupName"])
+	case err != nil:
+		return err
+	}
+
+	return f.kubeActions.Delete(ctx, doc.OpenShiftCluster, kind, namespace, name)
+}
+
+func (f *frontend) postAdminKubernetesObjects(w http.ResponseWriter, r *http.Request) {
+	ctx := r.Context()
+	log := ctx.Value(middleware.ContextKeyLog).(*logrus.Entry)
+	r.URL.Path = filepath.Dir(r.URL.Path)
+
+	body := r.Context().Value(middleware.ContextKeyBody).([]byte)
+	if len(body) == 0 || !json.Valid(body) {
+		api.WriteError(w, http.StatusBadRequest, api.CloudErrorCodeInvalidRequestContent, "", "The request content was invalid and could not be deserialized.")
+		return
+	}
+
+	err := f._postAdminKubernetesObjects(ctx, r)
+
+	reply(log, w, nil, nil, err)
+}
+
+func (f *frontend) _postAdminKubernetesObjects(ctx context.Context, r *http.Request) error {
+	body := r.Context().Value(middleware.ContextKeyBody).([]byte)
+	vars := mux.Vars(r)
+
+	resourceID := strings.TrimPrefix(r.URL.Path, "/admin")
+
+	doc, err := f.db.OpenShiftClusters.Get(ctx, resourceID)
+	switch {
+	case cosmosdb.IsErrorStatusCode(err, http.StatusNotFound):
+		return api.NewCloudError(http.StatusNotFound, api.CloudErrorCodeResourceNotFound, "", "The Resource '%s/%s' under resource group '%s' was not found.", vars["resourceType"], vars["resourceName"], vars["resourceGroupName"])
+	case err != nil:
+		return err
+	}
+
+	obj := &unstructured.Unstructured{}
+	err = obj.UnmarshalJSON(body)
+	if err != nil {
+		return api.NewCloudError(http.StatusBadRequest, api.CloudErrorCodeInvalidRequestContent, "", "The request content was invalid and could not be deserialized: %q.", err)
+	}
+
+	err = validateAdminKubernetesObjectsNonCustomer(r.Method, obj.GroupVersionKind().GroupKind().String(), obj.GetNamespace(), obj.GetName())
+	if err != nil {
+		return err
+	}
+
+	return f.kubeActions.CreateOrUpdate(ctx, doc.OpenShiftCluster, obj)
+}
+
+// rxKubernetesString is weaker than Kubernetes validation, but strong enough to
+// prevent mischief
+var rxKubernetesString = regexp.MustCompile(`(?i)^[-a-z0-9]{0,255}$`)
+
+func validateAdminKubernetesObjectsNonCustomer(method, kind, namespace, name string) error {
+	if namespace != "" &&
+		namespace != "default" &&
+		namespace != "openshift" &&
+		!strings.HasPrefix(string(namespace), "kube-") &&
+		!strings.HasPrefix(string(namespace), "openshift-") {
+		return api.NewCloudError(http.StatusBadRequest, api.CloudErrorCodeInvalidParameter, "", "The provided namespace '%s' is invalid.", namespace)
```

Comment:
Should really say access denied, as they are not invalid

Reply:
I wonder if this check should be in the .net client, as that is faster to update if in a pinch we need to fine tune this..

Reply:
I'm happy with it here. Updated.